common/sgx/eeid_verifier.c

```diff
@@ -0,0 +1,171 @@
+
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <stdlib.h>
+#include <string.h>
+
+#ifdef OE_BUILD_ENCLAVE
+#include <openenclave/enclave.h>
+#else
+#include <openenclave/host.h>
+#endif
+
+#include <openenclave/attestation/sgx/eeid_verifier.h>
+#include <openenclave/attestation/sgx/verifier.h>
+#include <openenclave/bits/eeid.h>
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/sgx/plugin.h>
+#include <openenclave/internal/trace.h>
+
+oe_result_t eeid_verifier_on_register(
+    oe_attestation_role_t* context,
+    const void* config_data,
+    size_t config_data_size)
+{
+    OE_UNUSED(context);
+    OE_UNUSED(config_data);
+    OE_UNUSED(config_data_size);
+
+    oe_result_t result = OE_UNEXPECTED;
+
+    oe_verifier_t* sgx_verify = oe_sgx_plugin_verifier();
+    result = oe_register_verifier(sgx_verify, NULL, 0);
+    return result == OE_ALREADY_EXISTS ? OE_OK : result;
+}
+
+oe_result_t eeid_verifier_on_unregister(oe_attestation_role_t* context)
+{
+    OE_UNUSED(context);
+    oe_result_t result = OE_UNEXPECTED;
+
+    oe_verifier_t* sgx_verify = oe_sgx_plugin_verifier();
+    result = oe_unregister_verifier(sgx_verify);
+    if (result != OE_OK && result != OE_NOT_FOUND)
+        return result;
+
+    return OE_OK;
+}
+
+typedef struct _header
+{
+    uint32_t version;
+    oe_uuid_t format_id;
+    uint64_t data_size;
+    uint8_t data[];
+} header_t;
+
+oe_result_t eeid_verify_evidence(
+    oe_verifier_t* context,
+    const uint8_t* evidence_buffer,
+    size_t evidence_buffer_size,
+    const uint8_t* endorsements_buffer,
+    size_t endorsements_buffer_size,
+    const oe_policy_t* policies,
+    size_t policies_size,
+    oe_claim_t** claims,
+    size_t* claims_length)
+{
+    OE_UNUSED(context);
+    OE_UNUSED(policies);
+    OE_UNUSED(policies_size);
+
+    oe_result_t result = OE_UNEXPECTED;
+
+    if (!evidence_buffer || evidence_buffer_size == 0 || !endorsements_buffer ||
+        endorsements_buffer_size == 0)
+        return OE_INVALID_PARAMETER;
+
+    eeid_evidence_t* eeide = (eeid_evidence_t*)evidence_buffer;
+    oe_claim_t* sgx_claims = NULL;
+    size_t sgx_claims_size = 0;
+    uint8_t* sgx_evidence_buffer = eeide->data;
+    size_t sgx_evidence_buffer_size = eeide->eeid_sz;
+    uint8_t* sgx_endorsements_buffer =
+        eeide->endorsements_sz ? eeide->data + eeide->evidence_sz : NULL;
```

Comment:
This line looks suspiciously like it could cause oe_verify_evidence to crash, but maybe the bug is in oe_verify_evidence?  The case I'm worried about is where eeide->evidence_sz is not a multiple of 4.  No code seems to test for that, and oe_verify_evidence assumes it can cast the endorsements buffer to oe_attestation_header_t and then access a 4-byte field, which would then be unaligned.

Reply:
That's a great catch, many thanks! I hadn't thought about alignment issues here at all. I'll do some digging to figure out where the fix should go.

Reply:
I think you're right; it's a bug on both ends. In the case of remote attestation we can't assume that all structures will be aligned, so we should fix it both in the plugin and in the infrastructure.
I added a host verification test (which was missing anyway) to trigger this in a controlled way, but I ended up fixing two related bugs, but not this one. Verification works without endorsements, but with endorsements it crashes in the guts of openssl while trying to load a certificate chain, even with aligned structs (or at least I think they are).